<a href="https://colab.research.google.com/github/jaquanjones/CS4661/blob/main/CS4661_KNN_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CS4661 - KNN Classification

## Objective:
  1. Determine increasing amount of nearest neighbots, K always improves the accuracy of predictions</br></br>
  2. Determine which two features provide the most accuracy independently in predictions</br></br>
  3. Determine which combination of two features provide the most accuracy in predictions</br></br>
  4. Determine if the two features providing the most accurate predictions independently are the same two features in the feature pair that provides the most accuracy

Reading the iris flower dataset from the following URL: https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv (Links to an external site.) and assign it to a Pandas DataFrame.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import pandas as pd



In [ ]:
# reading data from url
iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')

iris_df


sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

Splitting the dataset into testing and training sets with the following parameters: **test_size=0.4**, **random_state=6**

In [ ]:
from sklearn.model_selection import train_test_split

# set feature cols list
feature_cols = ['sepal_length','sepal_width','petal_length','petal_width'] 

# assign X as iris (not including species label column)
X = iris_df[feature_cols]

# display modified feature matrix
X


sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [ ]:
# print size/dimensions of Feature Matrix X
print(X.shape)

(150, 4)


In [ ]:
# label column values
y = iris_df['species']
y


0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

In [ ]:
print(y.shape)

(150,)


In [ ]:
# Randomly splitting the original dataset into training set and testing set
# test_size = 0.4, random_state=6

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

In [ ]:
# print size of training set
print(X_train.shape)
print(y_train.shape)

(90, 4)
(90,)


In [ ]:
# print size of testing set
print(X_test.shape)
print(y_test.shape)

(60, 4)
(60,)


In [ ]:
print('\t\t\t***X_train***\n')
print(X_train)
print('\n\t***y_train***\n')
print(y_train)

			***X_train***

     sepal_length  sepal_width  petal_length  petal_width
83            6.0          2.7           5.1          1.6
60            5.0          2.0           3.5          1.0
99            5.7          2.8           4.1          1.3
100           6.3          3.3           6.0          2.5
94            5.6          2.7           4.2          1.3
..            ...          ...           ...          ...
148           6.2          3.4           5.4          2.3
79            5.7          2.6           3.5          1.0
109           7.2          3.6           6.1          2.5
106           4.9          2.5           4.5          1.7
138           6.0          3.0           4.8          1.8

[90 rows x 4 columns]

	***y_train***

83     versicolor
60     versicolor
99     versicolor
100     virginica
94     versicolor
          ...    
148     virginica
79     versicolor
109     virginica
106     virginica
138     virginica
Name: species, Length: 90, dtype: object


In [ ]:
print('\t\t\t***X_test***\n')
print(X_test)
print('\n\t***y_test***\n')
print(y_test)

			***X_test***

     sepal_length  sepal_width  petal_length  petal_width
4             5.0          3.6           1.4          0.2
116           6.5          3.0           5.5          1.8
2             4.7          3.2           1.3          0.2
23            5.1          3.3           1.7          0.5
123           6.3          2.7           4.9          1.8
96            5.7          2.9           4.2          1.3
134           6.1          2.6           5.6          1.4
39            5.1          3.4           1.5          0.2
137           6.4          3.1           5.5          1.8
53            5.5          2.3           4.0          1.3
127           6.1          3.0           4.9          1.8
81            5.5          2.4           3.7          1.0
115           6.4          3.2           5.3          2.3
135           7.7          3.0           6.1          2.3
74            6.4          2.9           4.3          1.3
119           6.0          2.2           5.0          1

Creating an instance of KNN object with K=3, training it with the training set and testing it on the testing set. Then, calculating the accuracy of prediction made.


In [ ]:
# Instantiating another "object" of KNeighborsClassifier "class" with k=3:

k = 3
knn_3 = KNeighborsClassifier(n_neighbors=k)

In [ ]:
# "fit" method  of the object along with training dataset and labels to train the model.

knn_3.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

In [ ]:
# Using "predict" method of the *trained* object knn on testing set

y_predict_knn_3 = knn_3.predict(X_test) # generate predictions for testing set

print(y_predict_knn_3) # print predictions for testing set

['setosa' 'virginica' 'setosa' 'setosa' 'virginica' 'versicolor'
 'virginica' 'setosa' 'virginica' 'versicolor' 'virginica' 'versicolor'
 'virginica' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'virginica' 'setosa' 'setosa'
 'versicolor' 'versicolor' 'versicolor' 'virginica' 'setosa' 'versicolor'
 'setosa' 'versicolor' 'setosa' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'virginica' 'versicolor' 'setosa' 'setosa' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'versicolor' 'setosa'
 'setosa' 'versicolor' 'virginica' 'virginica' 'virginica' 'setosa'
 'virginica' 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica']


In [ ]:
# Evaluating accuracy of predictions by comparing the "predicted labels" for the Testing Set with its "actual labels"

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_predict_knn_3)

print(accuracy)

0.9666666666666667


In [ ]:
# Above code condensed as a method
# Returning accuracy of predictions for selected knn value

def accuracy_of_knn(k, X_training, y_training, X_testing, y_testing):

      # Create Instance of KNeighborsClassifier with parameter value of k
      selected_k = k 
      selected_knn = KNeighborsClassifier(n_neighbors=selected_k)
      
      # Train model with "fit" method along with training dataset and labels .
      selected_knn.fit(X_training, y_training)

      # Use "predict" method of the *trained* object selected_knn on testing set
      # Set 'y_prediction' to generated predictions for testing set
      y_predictions = selected_knn.predict(X_testing)

      # return overall accuracy of predicted labels vs actual values
      return accuracy_score(y_testing, y_predictions)

# Testing that same result received for knn 3
accuracy_of_knn(3, X_train, y_train, X_test, y_test)



0.9666666666666667

Repeating previous step with K=1, K=5, K=7, K=11, K=15, K=27, K=59. Evaluating if the accuracy always improves by increasing the value K.

**Observation**: Accuracy does not always improve by increasing the value of K. From the results it can be observed that the accuracy stedily increase starting from 1 to ending at 5. For values of K between 5 to 59, the accuracy value appears to decrease. Within the tested values of K, 5 resulted in the best accuracy while 59 resulted in the worst accuracy.

In [ ]:
# values of K to test
k_test = [1, 3, 5, 7, 11, 15, 27, 59]

results = {}

for k in k_test:
        results[k] = accuracy_of_knn(k, X_train, y_train, X_test, y_test)

for key, value in results.items():
        print(f'For K={key}, accuracy is: {value}\n')

For K=1, accuracy is: 0.95

For K=3, accuracy is: 0.9666666666666667

For K=5, accuracy is: 0.9833333333333333

For K=7, accuracy is: 0.9666666666666667

For K=11, accuracy is: 0.9666666666666667

For K=15, accuracy is: 0.9333333333333333

For K=27, accuracy is: 0.9166666666666666

For K=59, accuracy is: 0.8166666666666667



Experimenting making predictions based on only **ONE single feature**, to find the best single feature. Repeating previous experiment with K=3, four times (each time using only one of the 4 features), and computing the accuracy of each time and checking the accuracies. </br>Objective is to find which individual feature provide the best accuracy (or is the best feature) and which is the second best feature.

**Observation:** From the results obtained below it appears that 'petal_width' provides the best accuracy as an isolated feature with 0.95 accuracy, making it the the best feature. The second best feature would be 'petal_length', having the second highest accuracy of 0.93.

In [ ]:
# assign x1, x2, x3, x4 unique to each feature column of iris (not including species label column)
# ***input classifiers are expected to be 2D arrays***
# must reference as: iris_df[[feature_cols[column_index]]]

x1 = iris_df[[feature_cols[0]]]
x2 = iris_df[[feature_cols[1]]]
x3 = iris_df[[feature_cols[2]]]
x4 = iris_df[[feature_cols[3]]]

feature_list = [x1,x2, x3, x4]

In [ ]:
# method to train models dynamically for each new isolated feature matrix
def find_feature_accuracy(x_col, k):
        X_train, X_test, y_train, y_test = train_test_split(x_col, y, test_size=0.4, random_state=6)
        return accuracy_of_knn(k, X_train, y_train, X_test, y_test)

In [ ]:
# dictionary to map name of feature to accuracy of feature column data
feature_results = {}

for i in range(0,4):
        feature_results[feature_cols[i]] = find_feature_accuracy(feature_list[i], 3)

for key, value in feature_results.items():
        print(f'For feature \'{key}\', accuracy is: {value}\n')

For feature 'sepal_length', accuracy is: 0.7166666666666667

For feature 'sepal_width', accuracy is: 0.5666666666666667

For feature 'petal_length', accuracy is: 0.9333333333333333

For feature 'petal_width', accuracy is: 0.95



Now, I will be repeating the previous experiment, this time using two features. I will need to train, test, and evaluate my model for 6 different cases: using (1st and 2nd features), (1st and 3rd features), (1st and 4th  features), (2nd  and 3rd  features), (2nd and 4th features), (3rd and 4th  features).</br></br>Objective is to find which “feature pair” provides the best accuracy.

**Observation:** The feature pair providing the most accuracy is 'sepal_length' and 'petal_length', at 0.983 accuracy.

In [ ]:
# create pair list for indexing each unique pair of feature labels
pair_indexes = []
for i in range(0,3):
    for j in range(i+1,4):
        pair_indexes.append([i, j])

In [ ]:
# method to train models dynamically for each new feature pair matrix
def find_feature_pair_accuracy(index1, index2, k):
        feature_pair = iris_df[[feature_cols[index1], feature_cols[index2]]]
        X_train, X_test, y_train, y_test = train_test_split(feature_pair, y, test_size=0.4, random_state=6)
        return accuracy_of_knn(k, X_train, y_train, X_test, y_test)

In [ ]:
# pass pairs of features to feature pair accuracy method
accuracy_by_pair = lambda f1,f2, k: find_feature_pair_accuracy(f1, f2, k)

In [ ]:
# dictionary to map name of features to accuracy of feature columns data combination
feature_pair_results = []

for i in range(0,len(pair_indexes)):
        feature_pair_results.append([])
        feature_pair_results[i].append(feature_cols[pair_indexes[i][0]])
        feature_pair_results[i].append(feature_cols[pair_indexes[i][1]])
        feature_pair_results[i].append(find_feature_pair_accuracy(pair_indexes[i][0], pair_indexes[i][1], 3))

for data in feature_pair_results:
        print(f'For feature pair \'{data[0]}\' and \'{data[1]}\', accuracy is: {data[2]}\n')

For feature pair 'sepal_length' and 'sepal_width', accuracy is: 0.8166666666666667

For feature pair 'sepal_length' and 'petal_length', accuracy is: 0.9833333333333333

For feature pair 'sepal_length' and 'petal_width', accuracy is: 0.95

For feature pair 'sepal_width' and 'petal_length', accuracy is: 0.95

For feature pair 'sepal_width' and 'petal_width', accuracy is: 0.95

For feature pair 'petal_length' and 'petal_width', accuracy is: 0.9666666666666667



## Big Question: Does the “best feature pair” contain both “first best individual feature” and “second best individual feature”? Can it be concluded, that the “best two features” for classification are the first best feature along with the second best feature together?

---
**Final Observations:**</br>The results previously obtained showed 'petal_width' and 'petal_length' independently to be the best two features. On the other hand, the combination of features showing to be most accurate were 'sepal_length' and 'petal_length' and their combination was more accurate than both best individual features independently. Although  **'petal_width'** and **'petal_length'** were the best two features independently, they were not the same features pair that gave the best accuracy which were **'sepal_length'** and **'petal_length'**. Therefore, we cannot conclude that features with the best accuracy independently will also yield the best accuracy together as a pair.